### PR TITLE
refactor(shared-data): create protocol schema V6 

### LINKED ISSUE #8250: Mint JSON Protocol Schema v6
Ticket to nail down [protocol schema v6](https://opentrons.atlassian.net/wiki/spaces/RPDO/pages/694387662/Protocol+Schema+Version+6) 

Historical reference: https://app.zenhub.com/workspaces/app--ui-5f6248dba845b000139aa767/issues/opentrons/opentrons/6399

### PR DESCRIPTION
**Pull request by b-cooper**
# Overview

Mint the next iteration of the JSON Protocol Schema.  Many additions are meant to allow for this
schema's command spec to be widely applicable to any JSON serializable action.
This also includes a promotion of a few experimental features into the fold.

Closes #8250

#### Requested Additions To The Protocol Schema

- [x] a new command: `dispenseAirGap` 
- [x] rename `airGap` to `aspirateAirGap` for consistent naming
- [x] change `#/definitions/pipetteName` to an enum of all pipette names
- [x] tip positioning argument in x/y/z for all pipetting commands
- [x] liquid state represented outside of `designerApplication` (via `loadLiquid` command type)
- [x] "load" equipment commands `loadModules`, `loadPipettes`, `loadLabware` (will facilitate manual deck manipulation)
- [x] unique identifier for each command in list
- [x] add support for heater-shaker commands
    - [x] `heaterShakerModule/setTargetTemperature`
    - [x] `heaterShakerModule/setTargetShakeSpeed`
    - [x] `heaterShakerModule/awaitTemperature`
    - [x] `heaterShakerModule/awaitShakeSpeed` 
    - [x] `heaterShakerModule/deactivateHeater` 
- [x] flesh out `commandAnnotations` type to group atomic commands as “Steps“ (e.g. Transfer etc.)
- [x] add entry to robot model enum?
- [x] flexible `custom` commandType
-  ~a new command: `returnTipToTipRack`~ // We already support dropTip {pipette, labware, well}, which should allow for this
-  ~thermocycler semi configuration support~ // should be supportable through the `load_module` command 

#### Relevant Resources

- Initial RFC #6399
- [Confluence Doc](https://opentrons.atlassian.net/wiki/spaces/RPDO/pages/694387662/Protocol+Schema+Version+6)

# Changelog

- `loadPipette`, `loadModule`, `loadLabware` commands that are capable of defining the requirements to load an instance of a pipette, module, or labware, into a location. Location should be referenced by and id string which will have to be mapped to a slot or a raw coord elsewhere in the file (ala deck definition schema). 
- `loadLiquid` command that is capable of overwriting the tracked liquid volume contained in the specified wells of the specified labware

TODO: list out whole changelog here

# Review requests

Still in progress, nothing to review yet

# Risk assessment

low risk, though this schema update will have many implications for representing protocols as data across our ecosystem

### CODE REVIEW

File: shared-data/protocol/schemas/6.json [hunk 1]
```diff
@@ -0,0 +1,1017 @@
...
+
+    "offset": {
+      "description": "x, y, and z component of offset vector in mm",
+      "required": ["x", "y", "z"],
+      "properties": {
+        "x": { "type": "number" },
+        "y": { "type": "number" },
+        "z": { "type": "number" }
+      }
+    },
```
Comment (SyntaxColoring) [inline shared-data/protocol/schemas/6.json:44]: `offsetFromBottomMm` uses `mmOffset` instead of a raw `number`. Should `x`, `y`, and `z` here also be `mmOffset`s instead of `number`s?

I don't have a strong opinion, but I might just delete `mmOffset` in favor of using raw numbers everywhere?

File: shared-data/protocol/schemas/6.json [hunk 2]
```diff
@@ -0,0 +1,1017 @@
...
+
+    "pipetteAccessParams": {
```
Comment (SyntaxColoring) [inline shared-data/protocol/schemas/6.json:46]: Should these `*params` definitions forbid extra fields?

File: shared-data/protocol/schemas/6.json [hunk 3]
```diff
@@ -0,0 +1,1017 @@
...
+      }
+    },
+
+    "flowRate": {
+      "required": ["flowRate"],
+      "properties": {
+        "flowRate": {
+          "description": "Flow rate in uL/sec. Must be greater than 0",
+          "type": "number",
+          "minimum": 0
```
Comment (SyntaxColoring) [inline shared-data/protocol/schemas/6.json]: We can use `exclusiveMinimum` for this!

```suggestion
          "description": "Flow rate in uL/sec.",
          "type": "number",
          "exclusiveMinimum": 0
```

**Review by SyntaxColoring — COMMENTED** [review]
Just a few drive-by suggestions. 🚗

File: shared-data/protocol/schemas/6.json [hunk 4]
```diff
@@ -0,0 +1,1089 @@
...
+            "description": "The type of annotation (e.g. 'transfer', 'initialSetup')",
+            "type": "string"
+          },
+          "params": {
+            "description": "The parameters used to create this grouping of atomic commands",
+            "type": "object"
+          }
+        }
+      }
+    }
```
Comment (SyntaxColoring) [inline shared-data/protocol/schemas/6.json:1126]: * As written, it looks like this only allows a single level of nesting. (An annotation has a flat list of Protocol Engine commands.)

  I'm pretty sure we'll want to arbitrarily nest lists within lists.
  
  For example, a transfer may include many mixes, and each mix will have an aspirate and dispense.
  
  I also want to avoid us self-imposing unnecessary pressure for Protocol Engine to support more complex, non-atomic things as commands, just so we can fit them into this shape.

* With this way of grouping commands, do we have anything in mind for monitoring the run log as it evolves?

  I've been oversimplifying the run log in my brain to be a flat list of Protocol Engine commands, where it's not too hard to see a path for fitting it into [the current notifications and subscriptions](https://opentrons.atlassian.net/wiki/spaces/RPDO/pages/917110846/Notifications+and+subscriptions). We'd need a way for the client to receive jus the *new* command's ID so it doesn't have to re-fetch the entire list, but I have a few options in mind for that and it seems tractable.
  
  But with annotations a separate collection from the commands that they contain, this gets more complicated. What is the client monitoring? The list of annotations? The list of commands? Both?
  
  I'd like to avoid the client monitoring an `/annotations` endpoint concurrently with an independent `/commands` endpoint, because it seems likely to cause flickering problems when the client receives and processes a "command added" notification before the corresponding "`annotations["some-annotation-id"].commandIds` updated" notification.
  
  (If we don't have anything in mind yet, I'm happy for this to be something for me to figure out and make a proposal.)

* This is a nitpick, but I don't love the name `params` for this, nor the description that that object represents "the parameters used to create this grouping." I think something like `metadata` or `annotationInfo` might be more appropriate here? Something that connotes that this is a *description,* not a *prescription,* of what the grouping does.

  I wouldn't absolutely hate being overridden on this if we want to stay consistent with command models, though.
  
 -----
 
 (None of this should block merging as a draft, of course.)

Comment (b-cooper) [reply shared-data/protocol/schemas/6.json:1126]: As far as nesting, I assumed we could avoid recursively nesting things by just allowing different annotations to refer to the same command ids e.g. 
```
commandAnnotations: [
    {commandIds:  ['1', '2', '3'], annotationType: 'intialSetup' },
    {commandIds:  ['2', '3'], annotationType: 'moduleSetup' },
     ...
]
```
Shouldn't this be sufficient?

File: shared-data/protocol/schemas/6.json [hunk 5]
```diff
@@ -0,0 +1,1089 @@
...
+            }
+          }
+        },
+        {
+          "required": ["coordinates"],
+          "properties": {
+            "coordinates": {
+              "$ref": "#/definitions/offset",
+              "description": "X, Y, Z coordinates from the deck's origin point. (left-front-bottom of working space)"
+            }
```
Comment (shlokamin) [inline shared-data/protocol/schemas/6.json]: @mcous `deckLocation` can also be a coordinate

File: shared-data/protocol/schemas/6.json (hunk 4 above)
Comment (SyntaxColoring) [reply shared-data/protocol/schemas/6.json:1126]: > As far as nesting, I assumed we could avoid recursively nesting things by just allowing different annotations to refer to the same command ids...

Interesting. I hadn't considered that. That might work, but I think the simplicity that this wins us in the schema will be outweighed by the complexity that it will introduce elsewhere:

* If a client wants to build a UI to show a hierarchy of commands, it will be on the client to infer that hierarchy from the `commandAnnotations` list. This seems possible, but tricky.
    * How do you express a `fooComplexCommand` that contains a single `barComplexCommand` that contains four atomic commands, without making it ambiguous whether it's `foo` that contains `bar` or `bar` that contains `foo`?
        * We could adopt some convention where this is decided by the order in which the annotations appear in the list. But that doesn't feel quite right to me, personally. It's logically a question of structure, not a question of ordering, so the answer should be in the annotation structure, not the annotation order.
    * What do overlapping ranges, like `[1, 2], [2, 3]` mean? Is the robot server allowed to produce them? If so, how should the client interpret them?
    * Inferring the hierarchy seems like it will get slow when there are lots of annotations? I *think,* for every annotation, you'd have to check every other annotation to see if one logically forms a subset of the other.
* I don't currently have an answer for how the notifications system would account for this.

  As currently conceived, we can monitor individual resources for changes, and we can monitor whole *collections* of resources for changes, but I *don't* think we can monitor multiple arbitrarily-specified resources together.
  
  Continuing your example, say command `'4'` is added, and it's appended to both `initialSetup` and `moduleSetup`. The notifications system will currently issue two updates: one for the `initialSetup` resource, and one for the `moduleSetup` resource. Depending on the sequence in which the client receives them, things could temporarily look mis-nested, causing UI weirdness.

File: shared-data/protocol/types/schemaV6/command/setup.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+export type SetupCommand =
+  | { commandType: 'loadPipette'; params: LoadPipetteParams }
+  | { commandType: 'loadLabware'; params: LoadLabwareParams }
+  | { commandType: 'loadModule'; params: LoadModuleParams }
+  | { commandType: 'loadLiquid'; params: LoadLiquidParams }
+
+type DeckLocation =
```
Comment (mcous) [inline shared-data/protocol/types/schemaV6/command/setup.ts]: ProtocolEngine calls this a `LabwareLocation`, as in "valid locations for a labware", which I like better than `DeckLocation`. Deck location implies to me that the labware is on the deck, when it might, in fact, not be directly on the deck

File: shared-data/protocol/schemas/6.json [hunk 7]
```diff
@@ -0,0 +1,1089 @@
...
+                "required": ["labwareId", "location"],
+                "properties": {
+                  "labwareId": {
+                    "type": "string",
+                    "description": "Unique identifier of labware to load"
+                  },
+                  "location": {
+                    "$ref": "#/definitions/deckLocation",
+                    "description": "Physical destination location to load labware into (e.g. '1' || null)"
+                  }
```
Comment (SyntaxColoring) [inline shared-data/protocol/schemas/6.json]: This allows attempting to load a module on top of a module, or a module on top of a labware. Can we use more specific location types per command?

File: shared-data/protocol/schemas/6.json [hunk 8]
```diff
@@ -0,0 +1,1089 @@
...
+          "properties": {
+            "coordinates": {
+              "$ref": "#/definitions/offset",
+              "description": "X, Y, Z coordinates from the deck's origin point. (left-front-bottom of working space)"
+            }
+          }
+        },
+        {
+          "type": "null",
+          "description": "The null location corresponds to any location not tracked by the system (off-deck)"
```
Comment (SyntaxColoring) [inline shared-data/protocol/schemas/6.json]: How would this be used? If a command says it needs a location, what does it mean to pass it `null`?

File: shared-data/protocol/schemas/6.json [hunk 9]
```diff
@@ -0,0 +1,1089 @@
...
+                "required": ["pipetteId", "mount"],
+                "properties": {
+                  "pipetteId": {
+                    "type": "string",
+                    "description": "Unique identifier of pipette to load"
+                  },
+                  "mount": {
+                    "type": ["string", "null"],
+                    "description": "Unique identifier of physical destination location to load pipette into (e.g. 'left' || null)"
+                  }
```
Comment (SyntaxColoring) [inline shared-data/protocol/schemas/6.json]: What does it mean for the mount to be `null`?

File: shared-data/protocol/types/schemaV6/command/setup.ts (hunk 6 above)
Comment (b-cooper) [reply shared-data/protocol/types/schemaV6/command/setup.ts]: That makes sense to me, though this is a Location that doesn't only apply to Labware, so `Location` would then be more accurate to allow it to represent valid locations for Modules, Labware, etc. How would you feel about just `Location`?

File: shared-data/protocol/schemas/6.json (hunk 8 above)
Comment (b-cooper) [reply shared-data/protocol/schemas/6.json]: "The null location corresponds to any location not tracked by the system (off-deck)". This is particularly useful for specifying the destination of a labware (or other object) that is being move from some location that is tracked by the system (e.g. slot 1) to any location that the system doesn't track (e.g.  manually removed from the deck and placed on the bench next to the robot)

File: shared-data/protocol/schemas/6.json (hunk 7 above)
Comment (b-cooper) [reply shared-data/protocol/schemas/6.json]: I don't think we should be asserting those specifics at this level.  It may well be meaningful to stack a labware on top of another labware, and similarly, stacking a future module on top of another module.  We don't gain anything from being pedantic about it at the schema level. The actual execution/simulation layer should take into consideration the limitations of the system and the labwares/modules in question.

File: shared-data/protocol/schemas/6.json (hunk 9 above)
Comment (b-cooper) [reply shared-data/protocol/schemas/6.json]: It is the same as the `null` location for a module/labware.  Moving the instrument to a location that is not tracked by the system. Concretely this would capture the possibility of manually removing a pipette/instrument from the robot via a command that updates the robot state to be aware of the physical configuration change.

**Review by SyntaxColoring — APPROVED** [review]
I've looked over `6.json` and I'm happy for this to be merged as a draft. (Assuming we address the other comments and reviews somehow, even if "somehow" is just putting a pin in it until later.)

Here's a list of other discrepancies between PE and the schema that I noticed. This is probably non-exhaustive.

* `addLabwareDefinition` is present in PE but missing from the schema
    * I *think* the plan is to remove this from PE, right? So we'll have to update both PE and the schema for whatever `loadLabware` becomes because of that?
* The robot is still using `data` instead of `params` and I don't think we have a ticket to fix that.
* `moveToSlot` is missing from PE and we don't have a ticket for it. Will that be needed in the short term?
* `pickUpTip` and `dropTip` are separate commands in PE, but they're joined in the schema
* The schema's `wait` doesn't match PE's `pause`. The schema's `wait` is for both waiting for the user to resume, and waiting for an amount of time to pass. PE's `pause` is just for waiting for the user to resume.

File: shared-data/protocol/schemas/6.json (hunk 4 above)
Comment (b-cooper) [reply shared-data/protocol/schemas/6.json:1126]: In response to your first bullet point about inferring nested-ness. I imagine that you'd infer it in the same way that closures enclose logic in a script.  
In the example I gave:
```
commandAnnotations: [
    {commandIds:  ['1', '2', '3'], annotationType: 'intialSetup' },
    {commandIds:  ['2', '3'], annotationType: 'moduleSetup' },
     ...
]

________


{
  '1'; // only in the outer closure
  {
    '2'; // contained by both closures
    '3';
  }
}
```
because the `initalSetup`'s `commandIds` are a superset of the `moduleSetup`'s `commandIds` any client could infer that the `initialSetupComplexCommand` contains the `moduleSetupComplexCommand`.
 
 
 The `commandAnnotations` spec is purposefully loose. Because the `params` key corresponds to a vague `object` we could also always, cram whatever we want in there.

Re: "What do overlapping ranges, like [1, 2], [2, 3] mean?"

There are many possible reasons for wanting to annotate a sequence of commands.  As far as Annotating commands with `annotationType` strings that are solely used for organizing atomic commands sequences into larger and larger groupings of higher order commands, we can rely on the fact that there will never be overlapping ranges. These types of annotation groups will, by definition, be nested (supersets of the inner annotation's `commandIds`).

I like leaving the `commandAnnotations` spec more permissive. It will give us more flexibility to use it in other "command labeling" circumstances that we may not be able to predict now.  Ideally, we would maintain a separate list of enumerated `annotationTypes` that our system recognizes for delineating higher order commands etc.  and then leave the door open for anyone with more labelling needs to annotate any set of commands as they see fit (even if their ranges overlap).

File: shared-data/protocol/schemas/6.json (hunk 9 above)
Comment (SyntaxColoring) [reply shared-data/protocol/schemas/6.json]: Oh, so, effectively, this is a way for `loadPipette` to remove a pipette that a previous `loadPipette` had loaded?

But then, what does `params.pipetteId` mean here? It's a load name like `p300_single_gen2`, right? So then, if you have two `p300_single_gen2`s loaded (one on each mount), how does it know which one you want to remove?

File: shared-data/protocol/schemas/6.json (hunk 7 above)
Comment (SyntaxColoring) [reply shared-data/protocol/schemas/6.json]: > I don't think we should be asserting those specifics at this level.

Part of the problem/confusion might be that "this level" is really multiple things. If we're unifying Protocol Engine with protocol schema v6, then "this level" can mean:

* What Protocol Designer is allowed to output
* What other protocol authoring tools are allowed to output
* What the Opentrons App must be able to parse
* What the robot must be able to accept

I can only speak to "what the robot must be able to accept," but I absolutely *do* think that we should assert these specifics there.

> It may well be meaningful to stack a labware on top of another labware, and similarly, stacking a future module on top of another module

If that ever happens, we can specify it then and update the schema accordingly. Schema versioning seems like more effective future-proofing than trying to anticipate what we'll support and how it will work.

> We don't gain anything from being pedantic about it at the schema level.

We gain:

1. A contract, a single source of truth, for what is correct.
2. Static checks that Protocol Designer (or any 3rd-party authoring tool) cannot accidentally produce something that PE does not support, or that fundamentally doesn't make sense. (Especially since sometimes, we don't realize something doesn't make sense until we try to support it. 🙃)
3. Specification and documentation that's usable by a new member of the team, or a 3rd-party integrator, who doesn't know the details of what we currently support in `robot-server`.

File: shared-data/protocol/schemas/6.json (hunk 4 above)
Comment (mcous) [reply shared-data/protocol/schemas/6.json:1126]: I agree with @b-cooper on this one. A flat "commands" list with a permissive "annotations" structure is where I would like this to be

File: shared-data/protocol/types/schemaV6/command/setup.ts (hunk 6 above)
Comment (mcous) [reply shared-data/protocol/types/schemaV6/command/setup.ts]: I don't think they should be shared/generalized in this way. I think the individual "atoms" should be shared (e.g. `DeckSlot` > `{ slotId: str }`), but I think attempting to generalize the placement of "things" has gotten us into too much trouble in the past (see: `Placeable`).

I think there should be a `LabwareLocation` that applies to "valid places a labware can go", a `ModuleLocation` that applies to "valid places a module can go", etc. Start specific and explicit, generalize and loosen later

File: shared-data/protocol/schemas/6.json (hunk 7 above)
Comment (mcous) [reply shared-data/protocol/schemas/6.json]: Hard agree with @SyntaxColoring here (see my comment about the naming of `deckLocation`)

File: shared-data/protocol/schemas/6.json (hunk 8 above)
Comment (mcous) [reply shared-data/protocol/schemas/6.json]: I don't think this should be included right now. Nothing will use it, and it will make the engine logic / HTTP schema more confusing

File: shared-data/protocol/schemas/6.json (hunk 9 above)
Comment (mcous) [reply shared-data/protocol/schemas/6.json]: Same comment here about not including this at this time. It's easier to make a mandatory field nullable in the future than it is to go the other way around

File: shared-data/protocol/types/schemaV6/command/gantry.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
+export type GantryCommand =
+  | { commandType: 'moveToSlot'; params: MoveToSlotParams }
+  | { commandType: 'moveToWell'; params: MoveToWellParams }
+  | { commandType: 'moveToCoordinates'; params: MoveToCoordinatesParams }
+
+interface MoveToSlotParams {
+  pipetteId: string
+  slot: string
```
Comment (mcous) [inline shared-data/protocol/types/schemaV6/command/gantry.ts]: Some objects say `slot` and some say `slotId`. We should be consistent! I'm lightly in favor of `slot` (or maybe even `slotName`), but only very lightly

File: shared-data/protocol/types/schemaV6/command/gantry.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+interface MoveToWellParams {
+  pipetteId: string
+  labwareId: string
+  wellName: string
+  offset?: {
+    x: number
+    y: number
+    z: number
+  }
```
Comment (mcous) [inline shared-data/protocol/types/schemaV6/command/gantry.ts]: I think this should incorporate the ProtocolEngine's idea of a `WellOrigin`:

```python
class WellOrigin(str, Enum):
    """Origin of WellLocation offset."""

    TOP = "top"
    BOTTOM = "bottom"


class WellOffset(BaseModel):
    """An offset vector in (x, y, z)."""

    x: float = 0
    y: float = 0
    z: float = 0


class WellLocation(BaseModel):
    """A relative location in reference to a well's location."""

    origin: WellOrigin = WellOrigin.TOP
    offset: WellOffset = Field(default_factory=WellOffset)
```

File: shared-data/protocol/schemas/6.json (hunk 8 above)
Comment (b-cooper) [reply shared-data/protocol/schemas/6.json]: Actually, the `null` location for labware loading is very useful in the short term. Removing a labware from the deck mid protocol is something that would be quite useful with a number of workflows that take advantage of modules like the heater/shaker.  I'm fine with removing it as a formality for now, but I also think that it would be nice to agree that this is the functionality we'll need regardless.

Comment (mcous) [reply shared-data/protocol/schemas/6.json]: Definitely agree that this is functionality we want/need, but would still very much appreciate the formality of removing it for now! Would be cool to come at the problem all together when the feature is actually being implemented. As an example (but definitely not an actual proposal), what if we found out a `removeLabware` command could satisfy the requirement with some other benefit we hadn't thought of?

Comment (b-cooper) [reply shared-data/protocol/schemas/6.json]: Sounds good 👍 . I've removed it for now.

File: shared-data/protocol/schemas/6.json [hunk 12]
```diff
@@ -0,0 +1,1146 @@
...
+          "enum": ["OT-2 Standard", "OT-3 Standard"]
+        },
+        "deckId": {
+          "description": "Identifier of physical deck this protocol is written for. This should match a top-level key in shared-data/deck/definitions/2.json",
+          "type": "string"
+        }
+      }
+    },
+
+    "pipettes": {
```
Comment (mcous) [inline shared-data/protocol/schemas/6.json:196]: Shoot, missed this, but it applies to `pipettes`, `labwareDefinitions`, and `labware`: in our implemented endpoints, we tend to use lists rather than dictionaries for these collections. I prefer this for the server, because it allows a little more consistency in modeling collection and item resources. Historically, endpoints in our HTTP API that have returned dictionary collections have been hard to use and hard to extend.

I might, for example, want the top level `/run/:run_id` response to contain `labwareDefinitions` key with an array of `labwareDefinitionSummary`'s, and an endpoint `/run/:run_id/labwareDefintions/:definition_id`.

If we don't mind this mismatch, then we're all good, but if it makes sense to align this, I'd push pretty heavily for arrays given the ease at which we can choose to normalize however we need to on either end. It also ensures consistency of ordering in a way that a JSON dictionary, by definition, cannot

File: shared-data/protocol/schemas/6.json [hunk 13]
```diff
@@ -0,0 +1,1146 @@
...
+      "description": "The pipettes used in this protocol, keyed by a unique identifier (pipetteId)",
+      "additionalProperties": false,
+      "patternProperties": {
+        ".+": {
+          "description": "Fields describing an individual pipette",
+          "type": "object",
+          "required": ["name"],
```
Comment (mcous) [inline shared-data/protocol/schemas/6.json:203]: All `pipette`, `labware`, and `definition` resources on the server will have an `id`. I know as spec'd it's present as a key in the dictionary, but I think it should be in the value as well

File: shared-data/protocol/schemas/6.json [hunk 14]
```diff
@@ -0,0 +1,1146 @@
...
+            }
+          }
+        }
+      }
+    },
+
+    "liquids": {
+      "description": "All instances of liquid used in this protocol, keyed by a unique identifier (liquidId)",
+      "patternProperties": {
+        ".+": {
```
Comment (mcous) [inline shared-data/protocol/schemas/6.json:248]: Same list vs dictionary concern for liquids

File: shared-data/protocol/schemas/6.json [hunk 15]
```diff
@@ -0,0 +1,1146 @@
...
+      }
+    },
+
+    "commands": {
+      "description": "An array of command objects representing steps to be executed on the robot",
+      "type": "array",
+      "items": {
+        "anyOf": [
+          {
+            "description": "Aspirate Liquid / Dispense Liquid / Aspirate Air Gap / Dispense Air Aap",
```
Comment (mcous) [inline shared-data/protocol/schemas/6.json:290]: Can address this later, but internally on the server-side, every command is modeled individually. I think from a documentation standpoint, a union of individual commands is easier to manage (e.g. how do you write a good, concise "description" that has to cover multiple commands?).

I also think keeping them separate gives us more room to evolve each command. I wouldn't want to have to change the schema of dispense because we discovered we needed an new option for aspirates and we combined them early on when they happened to have the same payload.

TL;DR: explicit at the expense of repeating ourselves would make me happier

File: shared-data/protocol/schemas/6.json [hunk 16]
```diff
@@ -0,0 +1,1146 @@
...
+              "id": { "type": "string" },
+              "commandType": {
+                "enum": ["delay"]
+              },
+              "params": {
+                "type": "object",
+                "required": ["wait"],
+                "properties": {
+                  "wait": {
+                    "description": "either a number of seconds to wait (fractional values OK), or `true` to wait indefinitely until the user manually resumes the protocol",
```
Comment (mcous) [inline shared-data/protocol/schemas/6.json:496]: I would really like to ditch "`true` for wait indefinitely" in this spec. Fine to kick the can for merge

File: shared-data/protocol/schemas/6.json [hunk 17]
```diff
@@ -0,0 +1,1146 @@
...
+                      },
+                      {
+                        "required": ["coordinates"],
+                        "properties": {
+                          "coordinates": {
+                            "$ref": "#/definitions/offset",
+                            "description": "X, Y, Z coordinates from the deck's origin point. (left-front-bottom of working space)"
+                          }
+                        }
+                      }
```
Comment (mcous) [inline shared-data/protocol/schemas/6.json]: Unless we have something that is actively using these, I'd like to remove them for the same reason as `null` (so we can plan together as the feature is actually built)

File: shared-data/protocol/schemas/6.json [hunk 18]
```diff
@@ -0,0 +1,1146 @@
...
+                        "required": ["slotName"],
+                        "properties": {
+                          "slotName": {
+                            "description": "Unique string identifier that corresponds to a slot in the referenced deck definition's locations",
+                            "type": "string"
+                          }
+                        }
+                      },
+                      {
+                        "required": ["coordinates"],
```
Comment (mcous) [inline shared-data/protocol/schemas/6.json]: Loading labware and modules into arbitrary coordinates is cool, but I think requires a larger discussion so we can take the whole system into account for what this looks like

File: shared-data/protocol/schemas/6.json [hunk 19]
```diff
@@ -0,0 +1,1146 @@
...
+                      }
+                    ]
+                  }
+                }
+              }
+            }
+          },
+
+          {
+            "description": "Load Liquid: This command is only meant to make the Opentrons system aware of a liquid's location. It does not instruct any specific hardware to carry out the physical action of moving the liquid to the location.",
```
Comment (mcous) [inline shared-data/protocol/schemas/6.json:1050]: I think the payload of this command will take more work if we want this to work well in Python protocols. Like loadLabware and loadPipette in single-file Python protocols, these will also be the commands that create the resources themselves (rather than it existing in some pre-defined place)

**Review by mcous — COMMENTED** [review]
Nothing blocking merge, but definitely stuff I want addressed

File: shared-data/protocol/schemas/6.json (hunk 15 above)
Comment (SyntaxColoring) [reply shared-data/protocol/schemas/6.json:290]: Strongly agree.

File: shared-data/protocol/schemas/6.json (hunk 12 above)
Comment (b-cooper) [reply shared-data/protocol/schemas/6.json:196]: Sounds good to me. I'll add it to the follow-up list.

File: shared-data/protocol/schemas/6.json (hunk 16 above)
Comment (b-cooper) [reply shared-data/protocol/schemas/6.json:496]: yeah I'd love to ditch it too, I'll add to the follow-up list

File: shared-data/protocol/schemas/6.json (hunk 19 above)
Comment (b-cooper) [reply shared-data/protocol/schemas/6.json:1050]: That's a good call. We should have a discussion later on about how to split up "provisioning" and "placement".

**Comment by b-cooper** [comment]
All non blocking comments in this PR that include things that need to be addressed before this schema is frozen should be captured in this follow-up ticket. https://app.zenhub.com/workspaces/app--ui-5f6248dba845b000139aa767/issues/opentrons/opentrons/8583
